### PR TITLE
Big zone free and used lists

### PR DESCRIPTION
**Pull request by struct**
All chunk sizes in IsoAlloc are a power of 2. This can result in wasting memory in some cases (e.g. you request 2049 bytes and you get a chunk returned that 4096 bytes in size). This is fine at small sizes but at larger sizes it wastes a lot memory. This is particularly true if you request 32769 bytes as it will result in a 65535 chunk being returned. This PR introduces some changes to big zones that makes their size configurable via `SMALL_SZ_MAX`. Whatever this value is set to in `conf.h` is where big zones start to be used. This PR also introduces used and free lists for big zones and an additional mutex to reduce lock contention when iterating either list. There are still opportunities to introduce optimizations such as size specific lists and caches but that will come next.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -193,6 +193,16 @@ SIGNAL_HANDLER = -DSIGNAL_HANDLER=0
...
 
+## Enables guard pages around big zone meta data. Big zones
+## are zones with a size greater than SMALL_SZ_MAX as defined
+## in conf.h. Disabling this will save 8192 bytes per big
+## zone allocation
+BIG_ZONE_META_DATA_GUARD = -DBIG_ZONE_META_DATA_GUARD=1
+
+## Ensures all big zone user data pages are marked PROT_NONE
+## while on the free list to catch UAF
+PROTECT_FREE_BIG_ZONES = -DPROTECT_FREE_BIG_ZONES=0
```
Comment (jvoisin) [inline Makefile:204]: Why is this disabled by default? It should be documented :)

File: Makefile [hunk 2]
```diff
@@ -193,6 +193,16 @@ SIGNAL_HANDLER = -DSIGNAL_HANDLER=0
 ## supports Top Byte Ignore (TBI) then you want to enable this
 ARM_TBI = 0
 
+## Enables guard pages around big zone meta data. Big zones
+## are zones with a size greater than SMALL_SZ_MAX as defined
+## in conf.h. Disabling this will save 8192 bytes per big
```
Comment (jvoisin) [inline Makefile:198]: Why `8192` bytes?

File: PERFORMANCE.md [hunk 3]
```diff
@@ -12,16 +12,20 @@ Perhaps the most important optimization in IsoAlloc is the design choice to use
...
 
 Default zones for common sizes are created in the library constructor. This helps speed up allocations for long running programs. New zones are created on demand when needed but this will incur a small performance penalty in the allocation path.
 
+All chunk sizes are powers of 2 with a minimum value of `SMALLEST_CHUNK_SZ` (16 by default) and a maximum value of `SMALL_SZ_MAX` up to 65536 by default. This means zones can only be created for 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, and 65535. You can increase `SMALL_SZ_MAX` up to 131072. If you choose to change this value be mindful of the pages that it can waste (e.g. allocating a chunk of 16385 bytes will result in returning a chunk of 32768 bytes). Everything above `SMALL_SZ_MAX` is allocated on the big zone path which has a limitation of 4 GB and a size granularity that is only limited by page size alignment. Big zones have meta data allocated separately. Guard pages for this meta data can be enabled or disabled using `BIG_ZONE_META_DATA_GUARD`.
```
Comment (jvoisin) [inline PERFORMANCE.md:21]: Why `SMALLEST_CHUNK_SZ` instead of `SMALLEST_CHUNK_SIZE`?

File: README.md [hunk 4]
```diff
@@ -25,16 +23,16 @@ All user chunk pages and bitmap pages are surrounded by guard page allocations w
 
 If `DEBUG`, `LEAK_DETECTOR`, or `MEM_USAGE` are specified during compilation a memory leak and memory usage routine will be called from the destructor which will print useful information about the state of the heap at that time. These can also be invoked via the API, which is documented further below.
 
-* All allocations are 8 byte aligned.
+* All chunk sizes are a power of 2 and so are always aligned to 8.
```
Comment (jvoisin) [inline README.md:26]: Maybe phrase it like `are always 8 byte aligned.`

File: include/iso_alloc_internal.h [hunk 5]
```diff
@@ -285,11 +271,16 @@ extern uint32_t g_page_size_shift;
  * specific size request. */
 #define DEFAULT_ZONE_COUNT sizeof(default_zones) >> 3
 
-#define MEM_TAG_SIZE 1
+/* Each user allocation zone we make is 4mb in size.
```
Comment (jvoisin) [inline include/iso_alloc_internal.h:274]: I think you meant `MB` instead of `mb`,

File: PERFORMANCE.md (hunk 3 above)
Comment (struct) [reply PERFORMANCE.md:21]: No particular reason, its just always been the define name. Probably a good time to change it.

**Pull request merged**